toolchain/base/index_base.h

```diff
@@ -52,78 +53,32 @@ struct IndexBase : public IdBase {
   using IdBase::IdBase;
 };
 
-// Equality comparison for both IdBase and IndexBase.
-template <
-    typename IndexType,
-    typename std::enable_if_t<std::is_base_of_v<IdBase, IndexType>>* = nullptr>
+// Equality comparison for both IdBase and IndexBase, including the index type
+// on either the LHS or RHS with something convertible to it. Inequality
```

Comment:
As you're adding "including the index type on either the LHS or RHS with something convertible to it", I'm having trouble understanding what it's trying to say. On my first reading, index's type is `int32_t`, and "it" means the same, so this is saying "including int32_t with something convertible to int32_t". Would something like the below work for you?

```suggestion
// Support equality comparison when one operand is a child of IdBase (including IndexBase) and the other operand is either the same type or convertible. Inequality
```

Note though, I'm also wondering if we really need the "convertible" half of this -- #3670